dns/miekgdns2/resolver.go

```diff
@@ -185,21 +210,71 @@ func (r *Resolver) query(ctx context.Context, servers []string, attempts int, na
 	return nil, fmt.Errorf("could not resolve %s: no servers returned a viable answer. Errs %s", name, merr.Err())
 }
 
-// loadConfiguration parses and returns a resolv.conf configuration and builds a list of
-// nameservers of the form "ip:port" for convenience. Returns an error if the file cannot
-// be loaded or is not syntactically valid.
-func (r *Resolver) loadConfiguration() ([]string, *dns.ClientConfig, error) {
+// loop periodically reloads configuration from resolv.conf until the resolver is stopped.
+func (r *Resolver) loop() {
+	ticker := time.NewTicker(r.reloadPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := r.loadConfig(); err != nil {
+				level.Warn(r.logger).Log("msg", "unable to reload resolv.conf, using old values", "path", r.confPath, "err", err)
+			}
+		case <-r.stop:
+			return
+		}
+	}
+}
+
+// loadConfig loads and updates configuration from the configured resolv.conf path,
+// closing connections to defunct servers that are no longer configured, and returning
+// an error if the configuration file couldn't be opened or parsed. If an error is
+// returned, connections are not closed and the stored configuration is not updated.
+func (r *Resolver) loadConfig() error {
 	conf, err := dns.ClientConfigFromFile(r.confPath)
 	if err != nil {
-		return nil, nil, fmt.Errorf("could not load resolv.conf: %w", err)
+		return fmt.Errorf("could not load %s: %w", r.confPath, err)
 	}
 
 	servers := make([]string, len(conf.Servers))
-	for i, nameserver := range conf.Servers {
-		servers[i] = net.JoinHostPort(nameserver, conf.Port)
+	for i, ip := range conf.Servers {
+		servers[i] = net.JoinHostPort(ip, conf.Port)
 	}
 
-	return servers, conf, nil
+	// Close connections to any servers that are no longer in resolv.conf.
+	r.client.Clean(servers)
```

Comment:
Should we do this after updating `r.conf` below? Otherwise we might end up cleaning up some servers that will be immediately recreated before we have a chance to update `r.conf`.

(They'll be cleaned up the next time `loadConfig()` runs, but this seems tidier.)

Reply:
Sure, I'll move this after the update.